### ISSUE TITLE
PHP 8.1: Deprecation notices in several classes II

### ISSUE THREAD

**Issue by alexgit2k** [Shardj/zf1-future#214]
After using shardj/zf1-future now on production we have discovered further deprecation notices beside the one in #212:

- Zend_Form: shardj/zf1-future/library/Zend/Form.php, line 2125
  `strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Form.php#L2125  
- Zend_Form_Element: shardj/zf1-future/library/Zend/Form/Element.php, line 1099
  `strtoupper(): Passing null to parameter #1 ($string) of type string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Form/Element.php#L1099
- Zend_Captcha_Image: shardj/zf1-future/library/Zend/Captcha/Image.php, lines 548, 549, 550, 551, 573
  `Implicit conversion from float 123.4567890 to int loses precision`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Captcha/Image.php#L548
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Captcha/Image.php#L549
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Captcha/Image.php#L550
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Captcha/Image.php#L551
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Captcha/Image.php#L573
- Zend_Db_Adapter_Pdo: shardj/zf1-future/library/Zend/Db/Adapter/Pdo/Abstract.php, line 296
  `PDO::quote(): Passing null to parameter #1 ($string) of type string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Db/Adapter/Pdo/Abstract.php#L296
- Zend_Db_Statement: shardj/zf1-future/library/Zend/Db/Statement.php, line 200
  `preg_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Db/Statement.php#L200
- Zend_Uri: shardj/zf1-future/library/Zend/Uri.php, line 99
  `explode(): Passing null to parameter #2 ($string) of type string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Uri.php#L99
- Zend_View_Abstract: shardj/zf1-future/library/Zend/View/Abstract.php, line 904
  `htmlspecialchars(): Passing null to parameter #1 ($string) of type string is deprecated`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/View/Abstract.php#L903-L904

Beside the deprecations there is one warning:
- Zend_Form_Element: shardj/zf1-future/library/Zend/Form/Element.php, line 2280
  `foreach() argument must be of type array|object, string given`
  https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Form/Element.php#L2280

**Comment by IamKeck** [comment]
I found another one.

* Zend_Db_Adapter_Pdo_Pgsql: shardj/zf1-future/library/Zend/Db/Adapter/Pdo/Pgsql.php, line 214
`Error #8192 preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated`

https://github.com/Shardj/zf1-future/blob/master/library/Zend/Db/Adapter/Pdo/Pgsql.php#L214

**Comment by alexgit2k** [comment]
@Shardj Can you please reopen it. #218 only fixes the deprecation notices in Pgsql.php, but not the ones I reported.

**Comment by GustavR** [comment]
* Zend_Http_Response: shardj/zf1-future/library/Zend/Http/Response.php, lines 255, 270
`strtolower(): Passing null to parameter #1 ($string) of type string is deprecated`
https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Http/Response.php#L255
https://github.com/Shardj/zf1-future/blob/2436d37e1be56f6218342a9a86dae8e7ec89b804/library/Zend/Http/Response.php#L270

**Comment by icyz** [comment]
I found this: 
Deprecated: Return type of Zend_Soap_Client_Common::__doRequest($request, $location, $action, $version, $one_way = null) should either be compatible with SoapClient::__doRequest(string $request, string $location, string $action, int $version, bool $oneWay = false): ?string, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in `/Zend/Soap/Client/Common.php` on line 65 (error level: 8192)

https://github.com/Shardj/zf1-future/blob/ece10e31a53c8e40eb3faa321b44b277b4b974e9/library/Zend/Soap/Client/Common.php#L65

using `#[\ReturnTypeWillChange]`  temporary fix the problem

**Comment by Jimbolino** [comment]
most should be fixed in master:
https://github.com/Shardj/zf1-future/commit/25f2a1017544ad23c9af04b39152b73696e6ddcf
https://github.com/Shardj/zf1-future/commit/fdae0c27012e2fc47adab4c5cc8f74bb85d7f39f
https://github.com/Shardj/zf1-future/commit/55e259c09fe5383a20ea4f1c1476e0b8e58f7769
https://github.com/Shardj/zf1-future/commit/e9e24e7c047a7ff9feca4ebd89b02d9f985a2265

**Comment by michielfuture500** [comment]
Just ran into this one:

`Deprecated: file_get_contents(): Passing null to parameter #2 ($use_include_path) of type bool is deprecated in /opt/project/vendor/shardj/zf1-future/library/Zend/Translate/Adapter/Tmx.php on line 230`

**Comment by pascalheidmann** [comment]
Another one:
```
Exception ErrorException: "urldecode(): Passing null to parameter #1 ($string) of type string is deprecated" at /vendor/shardj/zf1-future/library/Zend/Rest/Route.php:194
```

**Comment by Jimbolino** [comment]
> Another one:
> 
> ```
> Exception ErrorException: "urldecode(): Passing null to parameter #1 ($string) of type string is deprecated" at /vendor/shardj/zf1-future/library/Zend/Rest/Route.php:194
> ```

https://github.com/Shardj/zf1-future/commit/52181d95eafe0e25d40f812f02745e3a056f9549

**Comment by michielfuture500** [comment]
@Jimbolino Would it be possible to also fix the deprecation I mentioned about a month ago? It would be highly appreciated.

**Comment by Jimbolino** [comment]
> @Jimbolino Would it be possible to also fix the deprecation I mentioned about a month ago? It would be highly appreciated.

this one?
https://github.com/Shardj/zf1-future/commit/844f99647d388ac915b36737316282966cf6340d

**Comment by michielfuture500** [comment]
@Jimbolino Yes, my sincere apologies for not checking better. Thanks!

**Comment by topicode** [comment]
I have found some more:

- `trim(): Passing null to parameter #1 ($string) of type string is deprecated in file /var/www/source/vendor/shardj/zf1-future/library/Zend/Form/Decorator/Label.php line 305`
- `preg_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated in file /var/www/source/vendor/shardj/zf1-future/library/Zend/Filter/PregReplace.php line 171`
- `strtoupper(): Passing null to parameter #1 ($string) of type string is deprecated in file /var/www/source/vendor/shardj/zf1-future/library/Zend/Locale/Math.php line 146`

**Comment by Jimbolino** [comment]
> I have found some more:

https://github.com/Shardj/zf1-future/commit/fac62dda23485aca33eada86975ff46a8f2bdf79

**Comment by mpdude** [comment]
@Jimbolino Do you think it would be possible to get an (intermediate, bugfix?) release for all the fixes you applied to `master` in the last days? 

That would help me to push out an update to my production servers and see if I can find any other issues, while reducing log message clutter for things already fixed.

**Comment by Shardj** [comment]
I'll make one for you now @mpdude

**Comment by Shardj** [comment]
@mpdude https://github.com/Shardj/zf1-future/releases/tag/release-1.21.3

**Comment by alexgit2k** [comment]
I have only rushed through the commits, I see that the bugs for zf1-future/library/Zend/Captcha/Image.php have been fixed in eeb300da114039fa5cc4d4d8c5fa266dd147763e and all of the ones in the first message as stated in https://github.com/Shardj/zf1-future/issues/214#issuecomment-1160687810

**Comment by mpdude** [comment]
Thank you @Shardj!
